src/server/frontend_wayland/ext_image_capture_v1.cpp

```diff
@@ -0,0 +1,604 @@
+    }
+}
+
+void mf::ExtImageCopyCaptureSessionV1::create_change_notifier()
+{
+    auto callback = [this, weak_self=wayland::make_weak(this)](std::optional<geom::Rectangle> const& damage)
+        {
+            ctx->wayland_executor->spawn([weak_self, damage]()
+                {
+                    if (weak_self)
+                    {
+                        weak_self.value().apply_damage(damage);
+                    }
+                });
+        };
+    change_notifier = std::make_shared<ms::SceneChangeNotification>(
+        [callback](){ callback(std::nullopt); },
+        callback);
+    ctx->surface_stack->add_observer(change_notifier);
+}
+
+void mf::ExtImageCopyCaptureSessionV1::maybe_capture_frame()
+{
+    if (!current_frame || !current_frame.value().is_ready())
+    {
+        // Do nothing if there is no frame to capture to.
+        return;
+    }
+
+    if (!change_notifier)
+    {
+        // Create the change notifier the first time a client tries to
+        // capture a frame.
+        create_change_notifier();
+    }
+
+    auto const& output_config = output.value().current_config();
+    auto const buffer_size = output_config.modes[output_config.current_mode_index].size;
+    auto const output_space_area = output_config.extents();
+    auto const transform = output_config.transformation();
+    bool const overlay_cursor = (options & wayland::ImageCopyCaptureManagerV1::Options::paint_cursors) != 0;
+
+    switch (damage_amount)
+    {
+    case DamageAmount::none:
+        return;
+
+    case DamageAmount::partial:
+        current_frame.value().capture_frame(
+            output_space_area,
+            transform,
+            overlay_cursor,
+            buffer_size,
+            translate_and_scale(
+                output_space_damage,
+                output_space_area,
+                {{}, buffer_size}));
+        break;
+
+    case DamageAmount::full:
+        current_frame.value().capture_frame(
+            output_space_area,
+            transform,
+            overlay_cursor,
+            buffer_size,
+            {{}, buffer_size});
+        break;
+    }
+
+    damage_amount = DamageAmount::none;
+}
+
+void mf::ExtImageCopyCaptureSessionV1::create_frame(
+    wl_resource* new_resource)
+{
+    if (current_frame)
+    {
+        BOOST_THROW_EXCEPTION(wayland::ProtocolError(
+            resource, Error::duplicate_frame,
+            "A frame already exists for this session"));
+    }
+    current_frame = wayland::make_weak(
+        new ExtImageCopyCaptureFrameV1{new_resource, this, ctx});
+}
+
+mf::ExtImageCopyCaptureFrameV1::ExtImageCopyCaptureFrameV1(
+    wl_resource *resource,
+    ExtImageCopyCaptureSessionV1* session,
+    std::shared_ptr<ExtImageCaptureV1Ctx> const& ctx)
+    : wayland::ImageCopyCaptureFrameV1{resource, Version<1>()},
+      session{session},
+      ctx{ctx}
+{
+}
+
+void mf::ExtImageCopyCaptureFrameV1::attach_buffer(
+    [[maybe_unused]] wl_resource* buffer)
```

Comment:
```suggestion
    wl_resource* buffer)
```
We can drop the `maybe_unused` now that this is implemented :)